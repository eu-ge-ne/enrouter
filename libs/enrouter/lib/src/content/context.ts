import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
