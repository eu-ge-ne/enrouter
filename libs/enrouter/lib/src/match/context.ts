import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
