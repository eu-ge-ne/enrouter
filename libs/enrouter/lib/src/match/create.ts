import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { loadRoutes } from "#lib/route/load.js";

const log = logger("match/create");

export interface CreateMatchParams {
  routes: Route;
  location: string;
}

export async function createMatch({
  routes,
  location,
}: CreateMatchParams): Promise<Match | undefined> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await loadRoutes(matches.map((x) => x.route));

  trim(matches);
  link(matches);

  log("matched: %o", matches);

  return matches[0];
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route.path, match.location);

  matches.push(match);

  if (match.route.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}

function trim(matches: Match[]): void {
  const rootIndex = matches.findLastIndex((x) => x.route.elements._root);
  if (rootIndex >= 0) {
    matches.splice(0, rootIndex);
  }

  if (!matches.at(-1)?.isFull) {
    const voidIndex = matches.findLastIndex(
      ({
        route: {
          elements: { _void, __void },
        },
      }) => _void || __void,
    );

    if (voidIndex >= 0) {
      if (matches[voidIndex]?.route.elements.__void) {
        const match = matches[voidIndex];
        matches.splice(0, matches.length);
        matches.push(match);
      } else {
        matches.splice(voidIndex + 1);
      }
    }
  }
}

function link(matches: Match[]): void {
  const first = matches[0];
  const last = matches.at(-1);

  matches.forEach((x, i) => {
    x.first = first;
    x.next = matches[i + 1];
    x.last = last;
  });
}
