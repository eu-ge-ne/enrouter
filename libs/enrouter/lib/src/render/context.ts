import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    handler: {
      route: {
        path: "",
        mod: [],
        link: [[], []],
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
    location: "",
    isFull: false,
    params: {},
  },
});
