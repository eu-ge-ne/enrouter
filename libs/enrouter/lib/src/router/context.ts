import { createContext, useContext } from "react";

import type { Route } from "#lib/route/mod.js";

export interface TRouterContext {
  routes: Route;
  location: string;
  navigate: (to: string) => void;
  ctx: unknown;
}

const RouterContext = createContext<TRouterContext>({
  routes: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },
  location: "",
  navigate: () => undefined,
  ctx: undefined,
});

export const RouterProvider = RouterContext.Provider;

export function useRouter(): TRouterContext {
  return useContext(RouterContext);
}
