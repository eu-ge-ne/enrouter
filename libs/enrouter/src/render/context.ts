import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    params: {},
    handler: {
      route: {
        path: "",
        mod: [],
        link: { css: [], mod: [] },
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
  },
});
