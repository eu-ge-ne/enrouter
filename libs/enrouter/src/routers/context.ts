import { createContext } from "react";

import type { RouteHandler } from "#handlers/mod.js";

export interface TRouterContext {
  handlers: RouteHandler;
  location: string;
  navigate: (to: string) => void;
}

export const RouterContext = createContext<TRouterContext>({
  handlers: {
    route: {
      path: "",
      mod: [],
      link: { css: [], mod: [] },
    },
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
  },
  location: "",
  navigate: () => undefined,
});
