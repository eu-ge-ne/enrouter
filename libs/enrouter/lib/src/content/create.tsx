import type { ReactElement } from "react";

import type { Match } from "#lib/match/mod.js";
import type { Content } from "./mod.js";
import { logger } from "#lib/debug.js";
import { ContentProvider } from "./context.js";
import { NotFound } from "./notFound.js";

const log = logger("content/create");

export function createContent(matches: Match[]): ReactElement[] {
  log("Creating matched content");

  const content = matches.map(toContent);
  const elements = render(content);

  log("Matched content created");

  return elements;
}

function toContent(match: Match): Content {
  const content: Content = {
    match,
  };

  function wrapChildren([key, children]: [string, ReactElement]) {
    const el = (
      <ContentProvider key={key} value={content}>
        {children}
      </ContentProvider>
    );
    return [key, el];
  }

  // do not render if match.next !== undefined
  if (!match.next && match.route.elements.index) {
    content.index = Object.fromEntries(
      Object.entries(match.route.elements.index).map((entry) => {
        log(
          'Creating content from index component "%s" matched by "%s"',
          entry[0],
          match.route.path,
        );
        return wrapChildren(entry);
      }),
    );
  }

  if (match.route.elements.layout) {
    content.layout = Object.fromEntries(
      Object.entries(match.route.elements.layout).map((entry) => {
        log(
          'Creating content from layout component "%s" matched by "%s"',
          entry[0],
          match.route.path,
        );
        return wrapChildren(entry);
      }),
    );
  }

  return content;
}

function render(content: Content[]): ReactElement[] {
  // 404?
  if (!content.at(-1)?.match.isFull) {
    log("404: %o", content);

    const i = content.findLastIndex((x) => x.match.route.elements.notFound);
    if (i === -1) {
      return [<NotFound />];
    }

    content = content.slice(0, i + 1);

    const x = content[i]!;
    x.index = x.match.route.elements.notFound;
  }

  const last = content.at(-1);
  content.forEach((x, i) => {
    x.prev = content[i - 1];
    x.next = content[i + 1];
    x.last = last;
  });

  return Object.values(content[0]?.layout ?? {});
}
