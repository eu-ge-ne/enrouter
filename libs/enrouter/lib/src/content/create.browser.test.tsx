import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import { type TRouterContext, RouterProvider } from "#lib/router/context.js";
import type { Match } from "#lib/match/mod.js";
import { createContent } from "./create.js";

const wrapperId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={wrapperId}>{children}</div>
);

describe("createContent", () => {
  test("from 0 matches", async () => {
    const context: TRouterContext = {
      routes: {
        path: "",
        test: { keys: [], pattern: new RegExp("") },
        modules: [],
        loaded: false,
        elements: {},
      },
      location: "/",
      navigate: () => undefined,
      ctx: undefined,
    };

    const matches: Match[] = [];

    const screen = render(createContent(matches), {
      wrapper: ({ children }) => (
        <div data-testid={wrapperId}>
          <RouterProvider value={context}>{children}</RouterProvider>
        </div>
      ),
    });

    await expect.element(screen.getByTestId(wrapperId)).toBeVisible();
    expect(screen.container).toMatchSnapshot();
  });

  test("from 1 match with no elements", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: false,
          elements: {},
        },
        location: "/",
        isFull: true,
        params: {},
      },
    ];

    const screen = render(createContent(matches), { wrapper });

    await expect.element(screen.getByTestId(wrapperId)).toBeVisible();
    expect(screen.container).toMatchSnapshot();
  });

  test("from 1 match with layout elements", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            layout: {
              main: <div>Layout</div>,
            },
          },
        },
        location: "/",
        isFull: true,
        params: {},
      },
    ];

    const screen = render(createContent(matches), { wrapper });

    await expect.element(screen.getByTestId(wrapperId)).toBeVisible();
    expect(screen.container).toMatchSnapshot();
  });
});
