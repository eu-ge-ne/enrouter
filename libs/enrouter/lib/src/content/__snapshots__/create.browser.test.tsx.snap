// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createContent > from 0 matches 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    / is not found
  </div>
</div>
`;

exports[`createContent > from 1 match with layout elements 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      Layout
    </div>
  </div>
</div>
`;

exports[`createContent > from 1 match with no elements 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  />
</div>
`;
